zilliqa/src/consensus.rs

```diff
@@ -1777,4 +1841,53 @@ impl Consensus {
 
         Ok(())
     }
+
+    /// Deal with the fork to this block. The block is assumed to be valid to switch to.
+    /// This will make it so the block is ready to become the new head
+    fn deal_with_fork(&mut self, block: &Block) -> Result<()> {
+        // Start at the head block and recursively unwind the changes that have been made
+        // until the head block is the parent of this block.
+        while self.head_block().header.number >= block.header.number {
+            let head_block = self.head_block();
+            let parent_block = self
+                .get_block(&head_block.parent_hash())?
+                .ok_or_else(|| anyhow!("missing block parent when reverting blocks!"))?;
+
+            if head_block.header.view == 0 {
+                panic!("genesis block is not supposed to be reverted");
+            }
+
+            trace!("Reverting block {}", head_block);
+            // block store doesn't require anything, it will just hold blocks that may now be invalid
+            // block transactions need to be removed from self.transactions and re-injected
+            for tx_hash in &head_block.transactions {
+                let orig_tx = self.get_transaction_by_hash(*tx_hash).unwrap().unwrap();
+                let _ = self.transactions.remove(tx_hash.0);
+
+                // Put tx back in.
+                self.new_transaction(orig_tx).unwrap();
+
+                // block hash reverse index, remove tx hash too
+                let _ = self.block_hash_reverse_index.remove(tx_hash.0);
+            }
+
+            // TX receipts are indexed by block
+            let _ = self.transaction_receipts.remove(head_block.hash().0);
+
+            // State is easily set - must be to the parent block, though
+            trace!("Setting state to: {}", parent_block.state_root_hash());
+            self.state
+                .set_to_root(H256(parent_block.state_root_hash().0));
+
+            // Persistence - only need to update head block pointer as it should be impossible
```

Comment:
Oh yeah... wait, this logic is way more dangerous than I thought now you point it out.

Because quite early on in the flow you add the proposed block to the database. This changes the canonical_block_numbers mapping, if you were to accidentally refer by number rather than hash lower down. I think I have to put it after the safe block check at least.

Reply:
ah, it isn't incorrect technically, but for whatever reason, moving the add block until after the safe check breaks consensus votes. I will ideally fix it in a follow-up PR.

Reply:
Technically, canonical block numbers aren't finalized block numbers - it's just what's _currently_ considered to be the main chain. By definition, the highest two blocks in that mapping are non-finalized and therefore subject to change.

At least that's how it's currently implemented. We could change it to a finalized block mapping only. But I think a lot of our chain architecture uses a best-effort model where we pipeline block creation and optimistically assume that the latest blocks will be fine even before finalization. In any case it's not too different to querying for `Latest` or just generally for a non-finalized block by hash, in all these cases it might happen that that blocks ends up being reverted.

Reply:
Just a note: the current implementation is, I believe, incorrect in that it considers blocks finalized so long as it is not one of the top two. This isn't actually the correct definition of finalized, which is that a block has a two blocks higher than it, where the views increment sequentially (so it is all happy path). I think (not too sure tho) that there can be arbitrary chain lengths that have no finalized blocks.

I don't see a lot of merit in referring to or thinking about finalized blocks at this stage, though. 

And, yeah I agree, don't think it should be changed to finalized block mapping only.